### PR TITLE
Added csp service

### PR DESCRIPTION
**Pull request by demianbrecht**
This change allows for modification of CSP settings at runtime,
defaulting to configuration settings.

Addresses #1005.

### CODE REVIEW

File: warehouse/csp.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        try:
+            policy = request.find_service(name="csp")
+        except ValueError:
+            policy = collections.defaultdict(list)
+
+        # We don't want to apply our Content Security Policy to the debug
+        # toolbar, that's not part of our application and it doesn't work with
+        # our restrictive CSP.
+        if not request.path.startswith("/_debug_toolbar/") and policy:
+            resp.headers["Content-Security-Policy"] = _serialize(policy)
```
Comment (dstufft) [inline warehouse/csp.py]: This should be ``_serialize(policy).format(request=request)``. We use a slightly dynamic CSP in development based on the request attributes. So right now this CSP is:

```
$ curl -I http://192.168.99.100/
HTTP/1.1 200 OK
Server: gunicorn/19.4.5
Date: Sat, 05 Mar 2016 16:10:54 GMT
Connection: keep-alive
Content-Type: text/html; charset=UTF-8
ETag: "zzqeSeiFKitpuq9Sj56MPw"
Content-Length: 11602
Content-Security-Policy: connect-src 'self'; default-src 'none'; font-src 'self' fonts.gstatic.com; frame-ancestors 'none'; img-src 'self' {request.scheme}://{request.domain}:9000/ https://secure.gravatar.com; referrer origin-when-cross-origin; reflected-xss block; script-src 'self'; style-src 'self' fonts.googleapis.com
Vary: Accept-Encoding
```

Which we obviously don't want values like ``{request.scheme}://{request.domain}:9000/``.

File: .gitignore [hunk 2]
```diff
@@ -23,3 +23,5 @@ dev/example.sql
 warehouse/.commit
 warehouse/static/components
 warehouse/static/dist
+
+tags
```
Comment (dstufft) [inline .gitignore]: What's this about?

**Comment by dstufft** [comment]
Two comments, otherwise looks good to me.

File: warehouse/csp.py (hunk 1 above)
Comment (demianbrecht) [inline warehouse/csp.py]: Ah cool, will change.

File: .gitignore (hunk 2 above)
Comment (demianbrecht) [inline .gitignore]: I use per-project ctags. `tags` is the default output file.

**Comment by demianbrecht** [comment]
Comments addressed and added a test to guard against regression of the csp formatting for development purposes.

**Comment by dstufft** [comment]
Looks great, thanks!

**Pull request merged**